contracts/protocol/tokenization/PToken.sol

```diff
@@ -215,33 +235,58 @@ contract PToken is
         address target,
         uint256 amount,
         DataTypes.TimeLockParams calldata timeLockParams
-    ) external virtual override onlyPool {
-        if (timeLockParams.releaseTime != 0) {
-            ITimeLock timeLock = POOL.TIME_LOCK();
-            uint256[] memory amounts = new uint256[](1);
-            amounts[0] = amount;
-
-            timeLock.createAgreement(
-                DataTypes.AssetType.ERC20,
-                timeLockParams.actionType,
-                _underlyingAsset,
-                amounts,
-                target,
-                timeLockParams.releaseTime
-            );
-            target = address(timeLock);
-        }
-        IERC20(_underlyingAsset).safeTransfer(target, amount);
+    ) public virtual override onlyPool {
+        _sendToUserOrTimeLock(
+            timeLockParams,
+            POOL.TIME_LOCK(),
+            _underlyingAsset,
+            amount,
+            target
+        );
     }
 
     /// @inheritdoc IPToken
-    function handleRepayment(address user, uint256 amount)
-        external
-        virtual
-        override
-        onlyPool
-    {
-        // Intentionally left blank
+    function swapAndTransferUnderlyingTo(
+        address target,
+        DataTypes.TimeLockParams calldata timeLockParams,
+        DataTypes.SwapAdapter calldata swapAdapter,
+        bytes calldata swapPayload,
+        DataTypes.SwapInfo calldata swapInfo
+    ) public virtual override onlyPool returns (uint256 amount) {
+        IERC20(swapInfo.srcToken).safeApprove(
+            swapAdapter.router,
+            swapInfo.maxAmountIn
+        );
+
+        bytes memory returndata = Address.functionDelegateCall(
+            swapAdapter.adapter,
+            abi.encodeWithSelector(
+                ISwapAdapter.swap.selector,
+                swapAdapter.router,
+                swapPayload,
+                swapInfo.exactInput
+            )
+        );
+        assembly {
+            returndata := add(returndata, 0x40)
```

Comment:
I think we should decode `amount` in the `SwapAdapter`,  this `0x40` logic is so confuse. We should let the `SwapAdapter` user handle this kind of logic.